[#37] Firebase를 활용한 푸시 서비스

cyj199637 opened the pull request:
- Firebase를 활용하여 푸시 서비스 구현
- 푸시 메시지 전송을 비동기로 처리

cyj199637 committed:
Merge pull request #4 from f-lab-edu/feature/3

[#3] 회원가입 기능 구현

cyj199637 committed:
Merge pull request #6 from f-lab-edu/feature/5

[#5] 로그인/로그아웃

cyj199637 committed:
Merge pull request #8 from f-lab-edu/feature/7

[#7] 회원정보 수정/회원탈퇴

cyj199637 committed:
Merge pull request #10 from f-lab-edu/feature/9

[#9] 단위 테스트 작성

cyj199637 committed:
Merge pull request #12 from f-lab-edu/feature/11

[#11] 프로필 정보 수정

cyj199637 committed:
[#13] 피드백 반영

- ObjectMapper를 Bean으로 등록하여 한번만 생성되도록 수정

cyj199637 committed:
[#13] 피드백 반영

- ObjectMapper를 빈으로 등록

cyj199637 committed:
[#13] 피드백 반영

- ObjectMapper 대신 GenericJackson2JsonRedisSerializer로 직렬화

cyj199637 committed:
[#13] RedisTemplate 설정

cyj199637 committed:
[#13] 피드백 반영

- CheckLoginAspect와 CurrentUserArgumentResolver에서 HttpSession에 관한 로직을 직접 다루지 않도록 분리
- Session에 User객체의 Id만 저장하도록 변경
- 위의 사항으로 인해 영향을 받는 사용자 정보 업데이트 기능 로직 변경

cyj199637 committed:
[#13] 피드백 반영

- LoginServie의 loginUser에서 User 객체 대신 userId 값만 받도록 수정

cyj199637 committed:
Merge pull request #14 from f-lab-edu/feature/13

[#13] Redis 세션 스토리지 연동

cyj199637 committed:
Merge pull request #24 from f-lab-edu/feature/23

[#23] 게시물 CRUD 캐싱 적용

cyj199637 committed:
Merge pull request #22 from f-lab-edu/feature/21

[#21] 게시물 수정/삭제

cyj199637 committed:
Merge pull request #20 from f-lab-edu/feature/19

[#19] 사용자 피드 조회

cyj199637 committed:
Merge pull request #18 from f-lab-edu/feature/17

[#17] 게시물 조회

cyj199637 committed:
Merge branch 'develop' into feature/15

cyj199637 committed:
Merge branch 'feature/15' into develop

cyj199637 committed:
Merge pull request #26 from f-lab-edu/feature/25

[#25] Redis 세션 저장소 - 캐시 저장소 간 분리

cyj199637 committed:
Merge pull request #28 from f-lab-edu/feature/27

[#27] 데이터베이스 Master/Slave 구성

cyj199637 committed:
Merge branch 'feature/25' into develop

cyj199637 committed:
[#37] Firebase를 활용한 푸시 서비스

- Firebase를 활용하여 푸시 서비스 구현
- 푸시 메시지 전송을 비동기로 처리

cyj199637 committed:
[#37] Firebase 푸시 서비스 수정

- 각 클래스마다 사용한 어노테이션에 대한 주석 추가
- AccessToken을 Redis에 저장하여 관리하도록 로직 변경

cyj199637 committed:
Merge branch 'develop' into feature/29

cyj199637 committed:
Merge branch 'feature/29' into feature/31

cyj199637 committed:
Merge branch 'feature/31' into feature/37

cyj199637 committed:
[#37] Firebase 푸시 서비스 수정

- 로그인 / 로그아웃에 AccessToken 등록 / 삭제 로직 추가

cyj199637 committed:
[#37] Firebase 푸시 서비스 빠트린 주석 추가 및 테스트 코드 수정

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/AsyncConfig.java:
```diff
@@ -0,0 +1,34 @@
+: @Async 어노테이션을 사용하면 Default로 SimpleAsyncTaskExecutor가 생성되어 사용된다.
+그러나 이는 어떤 스레드도 재사용하지 않고 호출마다 새로운 스레드를 생성하기 때문에 요청이 많은
+경우, 성능이 저하된다. 따라서, 새로운 스레드 풀을 생성하고 이를 프로젝트 특성에 맞게 튜닝한 후
+빈으로 등록한다.
```
👍

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/AsyncConfig.java:
```diff
@@ -0,0 +1,34 @@
+    public Executor taskExecutor() {
+        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+        executor.setCorePoolSize(3);
+        executor.setMaxPoolSize(50);
```
core pool size와 max pool size를 다르게 설정하신 이유가 있으실까요?

cyj199637 replied:
처리해야 할 요청이 많지 않은데도 최대 스레드를 유지하는 것은 불필요하게 리소스를 차지한다고 생각이 들어서 두 설정값을 다르게 정했습니다. 각각의 값은 아직 성능 테스트 전이라 임의로 설정했습니다!

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/AsyncConfig.java:
```diff
@@ -0,0 +1,34 @@
+        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+        executor.setCorePoolSize(3);
+        executor.setMaxPoolSize(50);
+        executor.setQueueCapacity(300);
```
1. `QueueCapacity`는 어떤 역할을 할까요?
2. 이 크기에 따라 각각 어떤 장단점이 있을까요?

cyj199637 replied:
1. QueueCapacity는 요청받은 작업을 작업 큐에 저장할 최대 갯수로 이 크기보다 많은 작업이 큐에 들어와야 poolSize가 하나씩 증가합니다.
2. 작게 잡으면 빠르게 새로운 스레드를 생성할 수 있으므로 작업 큐에서 대기하는 시간이 짧아지고 빠르게 처리될 수 있습니다. 그러나 요청 작업을 감당하기 위해 threadPool 크기가 커야 하므로 CPU를 많이 사용하고, 스레드 수가 증가하여 컨텍스트 스위칭에 대한 오버헤드가 증가하여 이로 인한 작업 처리 능력이 떨어질 수도 있습니다.
크게 잡으면 threadPool 크기를 작게 잡을 수 있으므로 이로 인해 CPU를 적게 사용하고, 컨텍스트 스위칭에 대한 오버헤드가 최소화될 수 있습니다. 그러나 Poolsize가 증가하지 않고 계속 CorePoolSize의 스레드만 사용하여 작업 처리 능력이 떨어질 수도 있습니다.

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/AsyncConfig.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Bean
+    public Executor taskExecutor() {
+        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
```
자바에서는 이런 스레드풀에 대한 팩토리 메소드를 제공합니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/FirebaseConfig.java:
```diff
@@ -0,0 +1,40 @@
+                FirebaseApp.initializeApp(options);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
`e.printStackTrace();`는 절대 쓰지 말아야할 메소드 중에 하나입니다. 왜 그럴까요?

cyj199637 replied:
printStackTrace() 메소드를 호출하면 서버에서 스택 정보를 취합해야하고 또한 이는 최대 100개까지 프린트할 수 있기 때문에 서버에 많은 부하가 발생하게 됩니다. printStackTrace() 대신 로거를 적용하여 예외 로그를 남기도록 수정하겠습니다!

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/RedisConfig.java:
```diff
@@ -24,4 +27,15 @@ RedisConnectionFactory redisConnectionFactory() {
     }
+
+    @Bean
+    RedisTemplate<String, Object> redisTemplate() {
```
이렇게 정의하면 spring default에 있는 `redisTemplate` bean을 대체하게 되는데요. 의도하신 내용일까요??

cyj199637 replied:
네! RedisTemplate이 default로 사용되는 serializer가 JdkSerializationRedisSerializer라고 공식 document에 쓰여져 있었습니다. JdkSerializationRedisSerializer로 직렬화하면 바이너리 값으로 저장되기 때문에 조회하는데 어려움이 있다고 판단했습니다. 그래서 key는 StringRedisSerializer로 value는 json 형태로 저장하기 위해 GenericJackson2JsonRedisSerializer를 사용하도록 재정의했습니다.

f-lab-dev commented on src/main/java/me/liiot/snsserver/enumeration/PushType.java:
```diff
@@ -0,0 +1,17 @@
+@Getter
+public enum PushType {
+
+    FOLLOWING("FOLLOWING", "%s님이 회원님의 계정을 팔로우하기 시작했습니다.");
```
1. 이 문구는 타입을 정의하는데 메세지도 같이 있습니다. 다른 성격의 데이터를 한 enum에 갖고있는게 아닐까요?
2. 만약 외국에 이 서비스를 런칭한다고 하면 이런 한글이 문제가 되지 않을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/AlarmService.java:
```diff
@@ -1,8 +1,21 @@
-    public void addAlarm(String userId, String targetId, AlarmType type);
+    private final AlarmMapper alarmMapper;
+
+    public void addAlarm(String userId, String targetId, AlarmType type) {
```
**알람을 추가하다**라는 이름보다는 **알람을 보내다**라는 이름이 더 어울릴 것 같습니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/FirebasePushService.java:
```diff
@@ -0,0 +1,73 @@
+
+    public void setToken(String userId) {
+
+        redisTemplate.opsForValue().set(userId, getAccessToken());
```
만료시간을 지정해주지 않는 것 같은데 그러면 메모리가 가득 차버리지 않을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/FirebasePushService.java:
```diff
@@ -0,0 +1,73 @@
+        redisTemplate.delete(userId);
+    }
+
+    @Async(value = "taskExecutor")
```
아래에서 이미 async로 전송하고 있는데 `@Async`를 다시 붙여줄 이유가 있을까요?

cyj199637 replied:
메소드를 비동기로 처리하려면 무조건 `@Async` 가 필요하다고 착각하였습니다. 수정하겠습니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/FirebasePushService.java:
```diff
@@ -0,0 +1,73 @@
+            googleCredential.refreshToken();
+            return googleCredential.getAccessToken();
+        } catch (IOException e) {
+            e.printStackTrace();
```
여기도 사용하지 말아야할 메소드가 있네요~

cyj199637 committed:
[#37] 피드백 반영

- MessageSource와 LocaleResolver를 이용하여 사용자의 위치에 따라 푸시 메시지의 언어가 다르게 나오도록 수정
- Redis에 저장되는 토큰 만료시간 설정
- 팩토리 메소드를 사용하여 스레드풀 생성

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/FirebasePushService.java:
```diff
@@ -41,18 +47,18 @@ public void deleteToken(String userId) {
         try {
-            FirebaseMessaging.getInstance().sendAsync(message).get();
-        } catch (ExecutionException | InterruptedException e) {
+            FirebaseMessaging.getInstance().send(message);
```
여기를 동기적인 방식으로 바꾸셨네요. 차라리 `@Async`를 빼는게 낫지 않을까요?? 이미 파이어베이스에서 비동기로 제공하고있기 때문에 `@Async`를 사용할 필요는 없어보입니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/WebConfig.java:
```diff
@@ -18,4 +22,12 @@
     }
+
+    @Bean
+    public LocaleResolver localeResolver() {
```
이런 다국어지원은 따로 이슈를 빼서 작업하는게 좋지 않을까요?

cyj199637 committed:
[#37] 피드백 반영

- @Async 대신 sendAsync() 메소드로 비동기 처리

f-lab-dev reviewed (approved):
추후 이력서를 쓸 때 이러한 비동기를 공부한 경험에 대해 스토리를 풀어주면 좋을 것 같네요~
